Use R / Bioconductor for Sequence Analysis
==========================================

Fred Hutchinson Cancer Research Center, Seattle, WA<br />
6-9 April, 2015

Contact: Martin Morgan ([mtmorgan@fhcrc.org](mailto:mtmorgan@fhcrc.org))

This INTERMEDIATE course is designed for individuals comfortable with
using _R_, and with at least passing familiarity with
_Bioconductor_. It consists of approximately equal parts lecture and
practical sessions addressing use of _Bioconductor_ software for
analysis and comprehension of high-throughput sequence and related
data. Specific topics include use of central Bioconductor classes
(e.g., _GRanges_, _SummarizedExperiment_), RNASeq gene differential
expression, ChIP-seq and methylation work flows, approaches to
management and integrative analysis of diverse high-throughput data
types, and strategies for working with large data. Participants are
required to bring a laptop with wireless internet access and a modern
version of the Chrome or Safari web browser.

Schedule (tentative)
--------------------

Day 1 (9:00 - 12:30; 1:30 - 5:00)

- Working with Genomic Ranges and other _Bioconductor_ data structures
  (e.g., in the
  [GenomicRanges](http://bioconductor.org/packages/devel/bioc/html/GenomicRanges.html)
  package).
- RNA-Seq known gene differential expression with
  [DESeq2](http://bioconductor.org/packages/devel/bioc/html/DESeq2.html).
- Gene set enrichment and other 'down stream' analysis.

Day 2 (9:00 - 12:30; 1:30 - 5:00)

- ChIP-seq and regulatory work flows with the
  [csaw](http://bioconductor.org/packages/devel/bioc/html/csaw.html) and
  [minfi](http://bioconductor.org/packages/devel/bioc/html/minfi.html).
- Integrative data analysis -- emerging approaches
- Working with large data -- efficient, parallel, and cloud
  programming.
